### PR TITLE
Jest testing

### PR DESCRIPTION
**Pull request by didoo**
*Issue #125*

**Why this PR:**
This is a huge refactoring of all the tests suite. I am opening the PR as a way to **open the discussion** about the possibility to switch to Jest (instead of Mocha). The benefits of this switch are:
- have a tool that allows us to detect changes not only in the formatters as functions, but also in the templates and in general on the _actual_ generated files/output. 
- use it as an excuse to review all the tests, extend the coverage and the reliability of the tests, and allow us to refactor the code with more confidence
- also, Jest is pretty standard in the community, quite handy using the "watch" mode, and above all more Modern™ :)

screenshot_1740

**Description of changes:**
- converted all the tests from Mocha to Jest using [jest-codemods](https://github.com/skovhus/jest-codemods)
- updated all the tests declaration to use the [Jest API methods](https://jestjs.io/docs/en/api) + removed any dependency of Chai + fixed some tests that were conceptually wrong
- moved some of the files inside the test folder (created subfolders to mimic the "lib" folder + added a "__" prefix to the folders used to contain assets, fixtures, etc)
- added snapshot tests for all the formatters

**Notes**:
- the CI tests are failing because in local the Date.now() returns GMT while on the server returns UTC (to investigate how to fix/workaround it)

**Leftovers**:
- update the documentation (!important)
- remove from the package.json the dependencies which are not needed anymore
- update the npm command for the coverage (is still using mocha, I don't know if/how _Istanbul_ can use Jest)
- possibly other things? I am not an expert in testing, sorry, so I am surely missing something

**Possible improvements**:
- Folder structure: do we want to keep all the tests under the `__tests__` folder in the root of the project, or move the tests near their corresponding code that is being tested? in that case, where do we store the resources (eg. JSON, assets, config, etc) and where do we write the outputs?
- Review all the tests, to see if they actually test what they are saying (in a couple of cases, the tests were wrong) and add the missing ones (at least, create an issue with the list and add them little by little)
- Have a "complete" case of properties/tokens declarations (currently in `__tests__/__properties`) that covers all the possible cases and variants. This will improve not only the tests for the snapshots, but in general for all the test suite. 
- better naming convention for folders and files (__properties, __assets, etc) and for variables (eg. StyleDictionaryExtended): to be discussed and agreed together

Also, during the development I have moved the files in the "test" folder to have a more similar folder structure to the "lib" folder. In doing that, I have noticed that some modules are not tested:
- lib/utils/flattenProperties.js
- lib/register/template.js
- lib/transform/config.js
- lib/transform/object.js
I imagine there will be more. Probably having the tests near the code would help in detecting this.

**Small questions**:
- Do we need to keep the "Copyright 2017 Amazon.com" on the top of the tests? It's just a test, is it really needed in the test files? We don't have similar headers for the JSON files :)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by chazzmoney** [comment]
A couple links I was perusing based on this conversation;

https://stackshare.io/stackups/jasmine-vs-jest-vs-mocha
https://medium.com/airbnb-engineering/unlocking-test-performance-migrating-from-mocha-to-jest-2796c508ec50

**Comment by dbanksdesign** [comment]
For the copyright text, I'm pretty sure... I'm not a open-source expert, I've just been told to have that in there. We can't comment in JSON files because you can't have comments in a valid JSON file :/ I'll double check about having that copyright comment in tests and get back to you

**Comment by dbanksdesign** [comment]
And this is an awesome start. I know cleaning up tests is not glamorous, but as you point out we definitely have some holes in our testing and a lot of room for improvement.

**Comment by didoo** [comment]
Cool, so looks like it's a go, no? In that case, can you have a look at the open questions above and share your thoughts and personal preferences? So I can proceed and close the last details (an fix the problem with the mock date that breaks the snapshot tests in CI).

PS: re. the copyright text, no problem at all @dbanksdesign, don't spend time on it. It was just strange to see this kind of text in a test file :)

**Comment by dbanksdesign** [comment]
Unless @chazzmoney has any objections, I'm good to go. I'll take a look more closely and try to answer open questions this weekend.

**Comment by chazzmoney** [comment]
Sorry for the delay @didoo - will take a look at this tomorrow and respond shortly thereafter

**Comment by chazzmoney** [comment]
- For the date issue, we could switch the tests to Date.UTC. Or we could use moment.
- Definitely want to push into develop, and probably Danny or I should create a jest-testing branch to push this into.
- Do you want to do all this work yourself? I think we might need to collaborate on this. We could create that jest-test branch and then all create PRs against it.
- I'm not opinionated on the folder structure, but I do think it could be improved. I fully trust you and Danny on making sure it is easy to understand which items have tests / which do not, as well as which tests apply to which pieces.
- Same for naming convention for folders and files (__properties, __assets, etc) and for variables (eg. StyleDictionaryExtended)

Also, here is my suggested TODO list for completion, stolen from your original comment:
- update the documentation
- remove from the package.json the dependencies which are not needed anymore
- update the npm command for the coverage (is still using mocha, I don't know if/how Istanbul can use Jest)
- Fix incorrect tests (will need to review all existing tests)
- Add missing tests (lib/utils/flattenProperties.js, lib/register/template.js, lib/transform/config.js, lib/transform/object.js)
- Have a "complete" case of properties/tokens declarations (currently in __tests__/__properties) that covers all the possible cases and variants. This will improve not only the tests for the snapshots, but in general for all the test suite.

Thank you for putting all this together and being so thoughtful about it!

**Comment by didoo** [comment]
- Regarding `Date.UTC` OK, I'll give it a try (I've seen there is also a `.toISOString()`). I would use moment.js only as last resort :)
- Working together on it would be great 🙌
- OK on creating a `jest-test` branch and then all create PRs against it.
- Todo list: I see that there are already a couple of "projects", maybe we could create another one for this task (I see that there are "cards" that can be moved like on a trello board); also I see that you can create "task lists" https://help.github.com/articles/about-task-lists/

File: __tests__/formats/es6Constants.test.js [hunk 1]
```diff
@@ -48,12 +47,22 @@ var dictionary = {
...
+  // beforeEach(() => {
+  //   helpers.clearOutput();
+  // });
+
+  // afterEach(() => {
+  //   helpers.clearOutput();
+  // });
+
+  describe('es6Constants', () => {
+    it('should be a valid JS file', () => {
```
[the remark this answers is not in the document]
Comment (dbanksdesign) [reply __tests__/formats/es6Constants.test.js]: It looks like we can: https://github.com/facebook/jest#additional-configuration

**Comment by didoo** [comment]
@chazzmoney @dbanksdesign I may have found a decent solution to the UTC problem. One of you can please run the test locally on your machine (if you prefer, create another local repo, so you can throw it away after the test) and see if all the tests are passing also for you? Thanks

**Comment by dbanksdesign** [comment]
@didoo For me the snapshot tests are still failing, but what I'm getting is the time is different now:
```
    -  * Generated on Sat, 01 Jan 2000 00:00:00 GMT
    +  * Generated on Sat, 01 Jan 2000 08:00:00 GMT
```

**Comment by dbanksdesign** [comment]
Also created a project for all this stuff: https://github.com/amzn/style-dictionary/projects/4

File: package.json [hunk 2]
```diff
@@ -36,14 +36,28 @@
   ],
   "scripts": {
     "lint": "eslint index.js lib/**/*.js test/*.js test/**/*.js",
-    "test": "npm run lint && mocha --recursive -c",
+    "test": "jest --runInBand",
```
[the remark this answers is not in the document]
Comment (didoo) [reply package.json]: my bad. restored in an upcoming commit.

File: __tests__/formats/es6Constants.test.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (didoo) [reply __tests__/formats/es6Constants.test.js]: I don't know how to do it :)

File: __tests__/utils/resolveObject.test.js [hunk 3]
```diff
@@ -0,0 +1,251 @@
...
+      expect(test).toHaveProperty('b', 1);
+      expect(test).toHaveProperty('c', 1);
+      expect(test).toHaveProperty('d', 1);
+      expect(test).toHaveProperty('e', 1);
+      expect(test).toHaveProperty('f', 1);
+      expect(test).toHaveProperty('g', 1);
+    });
+
+    it(
+      'should handle deep nested pointers with string interpolation',
```
[the remark this answers is not in the document]
Comment (didoo) [reply __tests__/utils/resolveObject.test.js]: some leftovers of the automatic conversion using jest-codemods. I thought I have cleaned them all, but clearly not :) fixed in an upcoming commit

File: __tests__/buildAllPlatforms.test.js [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+  afterAll(() => {
+    helpers.clearOutput();
+  });
+
+  it('should work with json config', () => {
+    expect(helpers.fileExists('./__tests__/__output/web/_icons.css')).toBeTruthy();
+    expect(helpers.fileExists('./__tests__/__output/android/colors.xml')).toBeTruthy();
+  });
+
+  it('should work with js config', () => {
```
[the remark this answers is not in the document]
Comment (didoo) [reply __tests__/buildAllPlatforms.test.js]: yes you're right. I didn't noticed it was doing two different tests, one for a JSON and one for a JS config file. I have fixed it in an upcoming commit.

**Comment by didoo** [comment]
> what I'm getting is the time is different now

ah, dates and javascript! 😢

File: __tests__/formats/es6Constants.test.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (dbanksdesign) [reply __tests__/formats/es6Constants.test.js]: No worries, I'll add that to the project and take it on

File: __tests__/buildAllPlatforms.test.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (dbanksdesign) [reply __tests__/buildAllPlatforms.test.js]: Looks good to me

[the remark this answers is not in the document]
Comment (dbanksdesign) [reply __tests__/buildAllPlatforms.test.js]: Looks good to me!

**Comment by dbanksdesign** [comment]
I created a new branch on this repo called 'jest' that we can start working from. I can change the branch on this PR and merge it in to the newly created branch so we can get started if you want @didoo. From there we can start tackling the open items in the [project](https://github.com/amzn/style-dictionary/projects/4) and filing PRs against that branch. I'm thinking about opening some of these smaller tasks to [hacktoberfest](https://hacktoberfest.digitalocean.com/) because adding some unit tests should be pretty easy?

**Pull request merged**